src/http/request.cr

```diff
@@ -93,32 +93,119 @@ class HTTP::Request
+
+      # See if we can find \r\n (first \n, then \r before it)
+      index = peek.index('\n'.ord.to_u8)
+      # TODO: eventually increase the 4096 bytes limit
+      if index && index > 0 && index < 4096 && peek[index - 1] == '\r'.ord.to_u8
+        parts = parse_request_line(peek[0, index - 1])
+        io.skip(index + 1) # Must skip until after \r\n
+        return parts
+      end
+    end
+
     request_line = io.gets(4096, chomp: true)
-    return unless request_line
+    return BadRequest.new unless request_line
+
+    parse_request_line(request_line)
+  end
+
+  private def self.parse_request_line(line : String) : RequestLine | BadRequest
+    parse_request_line(line.to_slice)
+  end
 
-    parts = request_line.split
-    return BadRequest.new unless parts.size == 3
+  private def self.parse_request_line(slice : Bytes) : RequestLine | BadRequest
+    space_index = slice.index(' '.ord.to_u8)
 
-    method, resource, http_version = parts
+    # Oops, only a single part (should be three)
+    return BadRequest.new unless space_index
 
-    return BadRequest.new unless HTTP::SUPPORTED_VERSIONS.includes?(http_version)
+    subslice = slice[0...space_index]
 
-    HTTP.parse_headers_and_body(io) do |headers, body|
-      request = new method, resource, headers, body, http_version
+    # Optimization: see if it's one of the common methods
+    # (avoids a string allocation for these methods)
+    method = METHODS.find { |method| method.to_slice == subslice } ||
+             String.new(subslice)
 
-      if io.responds_to?(:remote_address)
-        request.remote_address = io.remote_address.try &.to_s
-      end
+    # Skip spaces.
+    # The RFC just mentions a single space but most servers allow multiple.
+    while space_index < slice.size && slice[space_index] === ' '
+      space_index += 1
+    end
+
+    # Oops, we only found the "method" part followed by spaces
+    return BadRequest.new if space_index == slice.size
+
+    next_space_index = slice.index(' '.ord.to_u8, offset: space_index)
+
+    # Oops, we only found two parts (should be three)
+    return BadRequest.new unless next_space_index
+
+    resource = String.new(slice[space_index...next_space_index])
+
+    # Skip spaces again
+    space_index = next_space_index
+    while space_index < slice.size && slice[space_index] === ' '
+      space_index += 1
+    end
+
+    next_space_index = slice.index(' '.ord.to_u8, offset: space_index) || slice.size
+
+    subslice = slice[space_index...next_space_index]
+
+    # Optimization: avoid allocating a string for common HTTP version
+    http_version = HTTP::SUPPORTED_VERSIONS.find { |version| version.to_slice == subslice }
+    return BadRequest.new unless http_version
 
-      return request
+    # Skip trailing spaces
+    space_index = next_space_index
+    while space_index < slice.size
+      # Oops, we find something else (more than three parts)
+      return BadRequest.new unless slice[space_index] == ' '.ord.to_u8
```

Comment:
Half the character comparisons use `===` the other half `==` + `ord.to_u8`. I'd stick with either one of them, but keep it homogenous.

Reply:
Good idea. I left them all like `==` + `ord.to_u8` for consistency because when calling `index` and others you have to pass an `UInt8`.